src/tests/test_pyproxy.py

```diff
@@ -970,3 +967,65 @@ def f(*args):
         fpy.destroy();
         """
     )
+
+
+def test_pyproxy_this1(selenium):
+    selenium.run_js(
+        """
+        let f = pyodide.runPython(`
+            x = 0
+            def f(self, x):
+                return getattr(self, x)
+            f
+        `);
+
+        let x = {};
+        x.f = f.captureThis();
+        x.a = 7;
+        assert(() => x.f("a") === 7 );
+        f.destroy();
+        """
+    )
+
+
+def test_pyproxy_this2(selenium):
+    selenium.run_js(
+        """
+        let g = pyodide.runPython(`
+            x = 0
+            from pyodide.ffi import to_js
+            def f(*args):
+                return to_js(args)
+            f
+        `);
+
+        let f = g.captureThis();
+        let fjs = function(...args){return [this, ...args];};
+
+        let f1 = f.bind(1);
```

Comment:
Oddly enough, it is allowed to put a `1` there and it makes the assertions easier. It's not very useful to do this outside of testing `bind`.

Reply:
Relatedly, in JavaScript, `let x = new Number(1);` and `let x = Number(1)` don't do the same thing. Just what exactly does `new Number(1);` make? In what situation would one want such a thing? The docs tell you:
1. they don't do the same thing
2. never use `new Number(1)`, just use `Number(1)`
but they don't explain why `new Number(1)` is part of the language.

Reply:
Just to be completely clear, after `f.bind(1)`, `this` will be equal to `new Number(1)`.

Reply:
What I am confused is: `assert(() => JSON.stringify(g3(-1, -2, -3, -4)) === JSON.stringify([1, 2, 3, 4, 6, 7, 8, -1, -2, -3, -4]));` isn't this a wrong behavior?

```js
let foo = function(a, b, c) {
  console.log("a: " + a);
  console.log("b: " + b);
  console.log("c: " + c);
}

let bar = foo.bind(1);
bar(2, 3)
// a: 2
// b: 3
// c: undefined
```

Reply:
Well suppose I have a Python function which I am going to call from JavaScript, and it needs to access this. The typical way for a Python function to be bound is it takes a self parameter as a first argument. So for this to be accessed idiomatically, it should be the first argument. This is problematic because this is always around and often has junk that we don't care about. So I added a captureThis() function that switches the proxy to provide this as the first argument to the function in the case that this is desired. Does this make any sense?

Reply:
The analogous Python code:
```py
def foo(*args):
   return args
bar = foo.__get__(1)
bar(2, 3) # returns (1,2,3)
```

Reply:
Thanks for the explanation! I think it make sense, though I am still not sure how PyProxy with this can be useful in JavaScript (Maybe because I am not very savvy on JS).

Anyway, thanks for your work! It LGTM. Probably we need a some documentation as the behavior of `PyProxy.bind()` is quite different compared to normal JS bind.

Reply:
Well it should be useful if a Python function is being used in a JavaScript context where it needs access to `this`...